finality-aleph/src/base_protocol/handler.rs

```diff
@@ -49,122 +151,90 @@ where
 
-        // Check slot constraints depending on the node's role and the connection's direction.
-        if is_inbound
-            && handshake.roles.is_full()
-            && self.num_full_in_peers >= self.max_full_in_peers
-        {
-            return Err(ConnectError::TooManyFullInboundPeers);
-        }
-        if !is_inbound
-            && handshake.roles.is_full()
-            && self.num_full_out_peers >= self.max_full_out_peers
-        {
-            return Err(ConnectError::TooManyFullOutboundPeers);
+        match self.is_reserved(&peer_id) || self.limiter.allowed(&peer) {
+            true => Ok(peer),
+            false => Err(ConnectError::too_many_peers(peer)),
         }
-        if handshake.roles.is_light() && self.num_light_peers >= self.max_light_peers {
-            return Err(ConnectError::TooManyLightPeers);
-        }
-
-        Ok(handshake.roles)
     }
 
+    /// Accept or reject a peer.
     pub fn on_peer_connect(
         &mut self,
         peer_id: PeerId,
         handshake: Vec<u8>,
-        is_inbound: bool,
+        direction: Direction,
     ) -> Result<(), ConnectError> {
-        let role = self.verify_connection(peer_id, handshake, is_inbound)?;
+        let peer = self.verify_connection(peer_id, handshake, direction)?;
 
-        self.peers.insert(peer_id, PeerInfo { role, is_inbound });
-
-        if self.reserved_nodes.contains(&peer_id) {
-            return Ok(());
+        if !self.is_reserved(&peer_id) {
+            self.limiter.add(&peer);
         }
 
-        // Assign a slot for the node depending on their role and the connection's direction.
-        if is_inbound && role.is_full() {
-            self.num_full_in_peers += 1;
-        } else if !is_inbound && role.is_full() {
-            self.num_full_out_peers += 1;
-        } else if role.is_light() {
-            self.num_light_peers += 1;
-        }
+        self.peers.insert(peer_id, peer);
 
         Ok(())
     }
 
+    /// Clean up a disconnected peer.
     pub fn on_peer_disconnect(&mut self, peer_id: PeerId) -> Result<(), DisconnectError> {
-        let info = self
+        let peer = self
             .peers
             .remove(&peer_id)
             .ok_or(DisconnectError::PeerWasNotConnected)?;
 
-        if self.reserved_nodes.contains(&peer_id) {
-            return Ok(());
-        }
-
-        // Free the slot of the node depending on their role and the connection's direction.
-        if info.is_inbound && info.role.is_full() {
-            self.num_full_in_peers.saturating_dec();
-        } else if !info.is_inbound && info.role.is_full() {
-            self.num_full_out_peers.saturating_dec();
-        } else if info.role.is_light() {
-            self.num_light_peers.saturating_dec();
+        if !self.is_reserved(&peer_id) {
+            self.limiter.remove(&peer)
         }
 
         Ok(())
     }
+
+    /// Checks whether an inbound peer would be accepted.
+    pub fn accept_inbound(
```

Comment:
As for this function:
* Seems we could live without it, it's only used to handle `ValidateInboundSubstream`
* But if you want to keep it, maybe it should be renamed to `verify_inbound_connection`
* Moreover it's kinda tedious how it converts the inner `Ok` value to `()`. It could have the same return type as `verify_connection`, and we could handle it in `ValidateInboundSubstream` with `Ok(_) => ...`.